packages/task/src/browser/task-terminal-widget-manager.ts

```diff
@@ -0,0 +1,126 @@
+/********************************************************************************
+ * Copyright (C) 2020 Ericsson and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { inject, injectable } from 'inversify';
+import { ApplicationShell, WidgetManager, WidgetOpenerOptions } from '@theia/core/lib/browser';
+import { TerminalWidget } from '@theia/terminal/lib/browser/base/terminal-widget';
+import { TerminalWidgetFactoryOptions, TERMINAL_WIDGET_FACTORY_ID } from '@theia/terminal/lib/browser/terminal-widget-impl';
+import { TerminalService } from '@theia/terminal/lib/browser/base/terminal-service';
+import { TaskConfiguration, PanelKind } from '../common';
+import { TaskDefinitionRegistry } from './task-definition-registry';
+
+export interface TaskTerminalWidgetOpenerOptions extends WidgetOpenerOptions, TerminalWidgetFactoryOptions {
+    taskPanel?: PanelKind;
+    taskConfig?: TaskConfiguration;
+}
+export namespace TaskTerminalWidgetOpenerOptions {
+    export function isDedicatedTerminal(options: TaskTerminalWidgetOpenerOptions): boolean {
+        return !!options.taskPanel && options.taskPanel === PanelKind.Dedicated;
+    }
+
+    export function isNewTerminal(options: TaskTerminalWidgetOpenerOptions): boolean {
+        return !!options.taskPanel && options.taskPanel === PanelKind.New;
+    }
+
+    export function isSharedTerminal(options: TaskTerminalWidgetOpenerOptions): boolean {
+        return options.taskPanel === undefined || options.taskPanel === PanelKind.Shared;
+    }
+}
+
+@injectable()
+export class TaskTerminalWidgetManager {
+
+    @inject(WidgetManager)
+    protected readonly widgetManager: WidgetManager;
+
+    @inject(ApplicationShell)
+    protected readonly shell: ApplicationShell;
+
+    @inject(TaskDefinitionRegistry)
+    protected readonly taskDefinitionRegistry: TaskDefinitionRegistry;
+
+    @inject(TerminalService)
+    protected readonly terminalService: TerminalService;
+
+    // Map indexed by terminal widget id
+    protected terminalWidgetMap: Map<string, { isDedicated: boolean, taskConfig?: TaskConfiguration, widget: TerminalWidget }> = new Map();
```

Comment:
It does not look good, we should not cache `TerminalWidget` instances, it will lead to memory leaks. `WidgetManager` should be used as a single source of the truth. In this case you should use `TerminalService` as a facade to it.

You can cache though terminal widget ids as a reference which can be resolved later against `TerminalService` (and `WidgetManager` eventually)